File: src/016ThreeSumClosest/016ThreeSumClosest.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}
func threeSumClosest(nums []int, target int) int {
	n := len(nums)
	if n < 3 {
		return 0
	}
	sort.Ints(nums)
	targetdiff := math.MaxInt32
	targetclosest := target
	for k := 0; k < n-2; k++ {
		i := k + 1
		j := n - 1
		target2 := target - nums[k]
		for i < j {
			if nums[i]+nums[j] == target2 {
				return target
			} else if nums[i]+nums[j] < target2 {
				if abs(nums[i]+nums[j]+nums[k]-target) < targetdiff {
					targetclosest = nums[i] + nums[j] + nums[k]
					targetdiff = abs(targetclosest - target)
				}
				i++
			} else {
				if abs(nums[i]+nums[j]+nums[k]-target) < targetdiff {
					targetclosest = nums[i] + nums[j] + nums[k]
					targetdiff = abs(targetclosest - target)
				}
				j--
			}
		}
	}
	return targetclosest
}

func main() {
	fmt.Printf("%d\n", threeSumClosest([]int{-1, 2, 1, -4, 3, 2, 43, 1, -34}, 1))
}

File: src/016ThreeSumClosest/016ThreeSumClosest.rs
impl Solution {
    pub fn three_sum_closest(nums: Vec<i32>, target: i32) -> i32 {
        let mut nums = nums;
        nums.sort();
        let mut sum = std::i32::MAX;
        for k in 0..nums.len() as i32 -2 {
            let target2 = target - nums[k as usize];
            let (mut i, mut j) = (k+1 as i32, nums.len() as i32 -1);
            while i < j {
                let sum3 = nums[i as usize] + nums[j as usize] - target2;
                if sum3 == 0 {
                    return target;
                }else if sum3.abs() < sum.abs() {
                    sum = sum3;
                }
                
                if sum3 < 0 {
                    i+=1;
                }else{
                    j-=1;
                }
            }            
        }
        
        return sum+target;
    }
}
